### DISCUSSION TITLE
Fixing Gears of War Lens Flare

### DISCUSSION THREAD

**Discussion by bomabomaboma** [xenia-canary/game-patches#6] [General]
Every Gears of War title has the same problem in Xenia: 
Bright light sources are rendered by the camera, regardless of whatever geometry is in the way. It can be incredibly distracting.

I've solved the issue by identified the variable that needs to be changed in the game's data files, but a patch would be so preferable.

GearGame\Config\Xbox360-GearEngine.ini
[SystemSettings]
LensFlare:False

Basically, LensFlare needs to be disabled. Barely affects image quality, solves the issue completely for all of the GOW games.

I've tried using Ghidra, but I don't know what I'm doing. Here's hoping @illusion0001 or @emoose does.

**Comment by emoose** [comment]
IIRC changing the `query_occlusion_fake_sample_count` setting in xenia.config.toml helps with lensflare problems in some other games, maybe worth testing it with GoW too? (I think changing to -1 or 0 usually helped)

Besides that I guess you could see if the LensFlare default could be patched in the XEX, or find whatever code is checking it.

**Reply by bomabomaboma** [reply]
Unfortunately, that setting didn't fix the problem. 
I ran it through Ghidra and there are a couple references to LensFlare, but again, sorry I don't know how to reverse engineer.

I want to learn. Do you have any resources to recommend?

**Reply by emoose** [reply]
Well I've mostly learnt from trial & error over the years, so don't really have much that I can point to that helped me personally (other than Chris Eagle's IDA Pro & Ghidra books, which are great for learning the apps themselves)

I did see since Ghidra's release theres been a lot more YouTube tutorials going into REing things, if that suits your learning style maybe try looking into those (stacksmashing has some good videos going into Ghidra, eg https://www.youtube.com/watch?v=fTGTnrgjuGA)

Reddit's /r/ReverseEngineering subreddit has some good writeups posted there from time to time too (eg. https://reversing.games/jekyll/update/2021/03/05/osrs-player-npc-list.html, which kind of seems similar to what might be needed here, finding code that makes use of a string etc)

**Reply by illusionyy** [reply]
Will have a look

**Reply by illusionyy** [reply]
Where can I even see this lens flare? @bomabomaboma Can you provide a screenshot and level to load?

**Reply by bomabomaboma** [reply]
> Where can I even see this lens flare? @bomabomaboma Can you provide a screenshot and level to load?

At the beginning of the GoW3 campaign you are on a ship. The bridge deck lamps bleed through geometry about 5-10 minutes into the campaign, and they show up over and over again in all these games. 
![gow3lensflares](https://user-images.githubusercontent.com/81534189/114324401-92b9cf00-9ade-11eb-9576-939c28e3b079.png)


This is even easier to diagnose. Immediately after the intro campaign video, you can see the issue in Gears of War: Judgment.
![gowjlensflare](https://user-images.githubusercontent.com/81534189/114324253-f7285e80-9add-11eb-8d3a-fe8f2beb6886.jpg)

Admittedly, I haven't played GoW 1 in Xenia, but 2/3/Judgment all have this same problem with LensFlares.
Sorry, I don't have an example from GoW 2, but trust me, it happens.

**Reply by illusionyy** [reply]
Thank you, will update you on the patch when it's done

**Reply by bomabomaboma** [reply]
BTW, I'm not sure which variable you changed to unlock vsync in any of these Unreal Engine 3 games. Obviously this applies to far more than just these silly-but-fun military shooters.

Setting "bSmoothFrameRate" from whatever game's equivalent of their engine.ini is, to FALSE will fix framepacing issues that come with unlocking the framerate. That was the case in earlier UE3 games even on PC, and it's absolutely the case with Xenia.

I was heavily involved in the script modding scenes of several UE3 games a decade ago.

Looking through Ghidra, I found the "bSmoothFrameRate" string in those .xex's, but I'm not sure how to follow the bread crumbs whilst using the Xenia debugger to find the right address to modify.

**Reply by illusionyy** [reply]
@bomabomaboma Pushed https://github.com/xenia-canary/game-patches/commit/d40a537b67c0381283284451f0d395e40a603f00

**Reply by illusionyy** [reply]
> I found the "bSmoothFrameRate" string in those .xex's, but I'm not sure how to follow the bread crumbs whilst using the Xenia debugger to find the right address to modify.

I already made patch that disables bSmoothFrameRate. Give it a go

**Reply by bomabomaboma** [reply]
Cheers, so that's what your FPS patches changed!

BTW, I've figured out how to fix the grass pop-in/LOD, and (mostly) the horse-calling issue in Red Dead Redemption via file modding. Unfortunately it involves changing many, many, many values everywhere. 

Maybe Triangle's planned new GPU improvements may help with that. If not, I'll log everything, write them down and share my findings.

**Reply by illusionyy** [reply]
I think converting into xex patch maybe doable but might be a waste of time due to new gpu improvement coming in the not too distant future. Might be fun to learn more about Rage engine though. Also have you discover anything related to graphics settings like motion blur?

**Reply by bomabomaboma** [reply]
So I played through two random act chapters in Gears 2 and couldn't see any evidence of improperly shadered lens flares. I swear I've witnessed it in the latter half of the game, but I can't find any instance. *shrug*

Gears 2 emulation has its own myriad of unique issues.

**Reply by FlintEastwood** [reply]
Hello boma and illusion0001,
You're really doing a great work on Xenia, I really appreciate this.

@bomabomaboma:
For GoW2, I remember a level in Landown, with a very deep Sunset. There you enter a tunnel, and the sun was shining heavy through all objects, when you turned around. That was very distracting.

I've seen your patches for the Lens Flare problems.
Unfortunatelly my Gears disks have different module hashes. (EU Pegi18 versions)
```
Title ID	Title			Module Hash		Media ID
4D53082D	Gears Of War 2		6A8DC97B386CB14C	51D281BE
4D5308AB	Gears Of War 3		A594C55977266D00	7554392D
```
I guess, it would be hard to adopt your patches, as I'm totally new to this.
Anyway, I found a tool to edit the coalesced_INT.bin/coalesced_DEU.bin and disabled the lensflares that way. 

I also found entries for ResX and ResY and I was wondering about, if the game could be natively set to 1920x1080. But that seems to be hardcoded in the XEX ?

And a bit Offtopic: Is there a method to merge the TU's directly into the game-files? Google was not really helpful.

**Reply by oSerenity** [reply]
@FlintEastwood Please inform us of the version number of the game since this may be a reason why you are getting a different hash

**Reply by FlintEastwood** [reply]
Gears 2 (shows in title-bar as v7.0):
```
Execution Id
  Media Id:           51D281BE
  Title Id:           4D53082D  (MS-2093)
  Savegame Id:        00000000
  Version:            v0.0.0.7
  Base Version:       v0.0.0.7
```
Gears 3 (shows in title-bar as v6.0):
```
Execution Id
  Media Id:           7554392D
  Title Id:           4D5308AB  (MS-2219)
  Savegame Id:        00000000
  Version:            v0.0.0.6
  Base Version:       v0.0.0.6
```
I found a video today, that shows how to merge TU's directly into the game data with velocity and XEXtool. Maybe I should try this?

**Reply by bomabomaboma** [reply]
@FlintEastwood Your base version of Gears 2 and Gears 3 are the same as my NTSC-U version for which my (and Illusion's) patches were based off of. -- Try appending the patch files with your hash and see if the patches are cross-compatible, because they often are region-agnostic, provided the base version is the same.

As for using title updates: Many Unreal Engine 3 games on the 360 can't be updated using the xextool method because the TU container files contain not only delta patches for the .xex but also delta patches for engine files. You'll probably end up getting a dirty disc error. It's worth a shot though. Some games will not be able to be patched until proper TU support is reintegrated into Canary.

And thank you for reporting back on the Gears 2 lens flare issue. I'll make a patch for it.

**Reply by FlintEastwood** [reply]
@bomabomaboma : You were right. The patches are working when I change the module hashes. Thank you. :)

A bit offtopic:
For Gears 2 I wanted to modify the coalesced_int.bin as I did for Gears 3 before, and the new md5 hash is correctly written to the XBox360TOC.txt, but Xenia-Canary complains about a "Disc read error". Maybe the md5 hashes are still saved somewhere else?

**Reply by bomabomaboma** [reply]
@FlintEastwood The base version of Gears 3, for whatever reason, was the only Gears game on 360 that didn't generate SHA tables and run a hash check at startup. Later TUs would amend that. The other games require a bit of an involved process to change the Coalesced files.

Decrypt and decompress default.xex using xextool, then open it in a hex editor.
![image](https://user-images.githubusercontent.com/81534189/153771192-41dd9ee7-67f4-4866-a643-ecf1ef92bbe2.png)
You'll need to enter in the SHA-1 of your newly edited file in hex. Bear in mind that this will change your hash, and you'll need to update the patch file on line 2.

This is not really a conversation related to Xenia patch making. If you have any further questions about it, please befriend me on Discord.

**Reply by FlintEastwood** [reply]
@bomabomaboma :👍 Thank you for explaining. Sorry, I didn't want to spam here about modifing Gears games.
The deep sun in Gears 2 Landown tunnel is really LensFlare related. I just wanted to check this, as it looks a bit different as the Gears3 Lensflare stuff.
![GoW2_1_LensFlaresOff](https://user-images.githubusercontent.com/19540104/153773047-0cd3bf1b-3a6e-4d60-8b12-9929bca6ba71.png)
![GoW2_1_LensFlaresOn](https://user-images.githubusercontent.com/19540104/153773055-d0cbdca7-aea4-40a1-ac2b-b54983110f8a.png)
![GoW2_2_LensFlaresOff](https://user-images.githubusercontent.com/19540104/153773059-9552e8d8-b020-471f-899e-679b110b9aa3.png)
![GoW2_2_LensFlaresOn](https://user-images.githubusercontent.com/19540104/153773064-63f16f88-60d3-4ca0-a660-0a03dafbd7ad.png)

**Reply by bomabomaboma** [reply]
@FlintEastwood I made a patch for it last night.
[https://github.com/xenia-canary/game-patches/blob/main/patches/4D53082D%20-%20Gears%20of%20War%202.patch](url)
